### DISCUSSION TITLE
Submit + save to the database in a wizard form before the final step.

### DISCUSSION THREAD

**Discussion by dave-mills** [filamentphp/filament#8731] [Help] [answered]
Hi all, 

TLDR: Is there any way to save a new data model after Step 1 of a multi-step Wizard form? So Step 1 is in the "create" operation, and then Steps 2 onward are editing the newly created model? 

Right now I have it setup so the create page *only* shows Step 1, and then after submitting it goes to the edit page, which shows all the steps - and I use the `getStartStep()` method to start at step 2. This is fine, but I'd really like to show the user that the other steps exist right from the start... 

### Why do I want to do this? 
I have a workflow that involves multiple steps:

- Step 1 - The user uploads an Excel file. 
- Steps 2 + 3 - Based on the file contents the user might need to add extra information. The information needed is not known until the file is uploaded and processed, and right now that processing happens when the model is saved. I'd rather not change this, as the processing involves creating a bunch of related records, and it's much simpler to do that when the record actually exists in the database. 

To me, a huge benefit of the "Wizard" layout is that users can see all the required steps in the process right from the start. But with my current setup, I can't get this benefit. I'm probably going to create my own custom components to let me have all the steps displayed without being active in the Create page, but before I go down that rabbit hole, I was wondering if:

a) I'm missing something that I can do with the built-in components.
b) I'm missing some other approach I can take that'd be cleaner!

Thanks!

**Comment by leandrocfe** [comment, marked as answer]
`->afterValidation()` ?

**Reply by dave-mills** [reply]
Good thought - I tried a few possibly naieve things like `$this->create()` in the afterValidation() - which silently failed and stopped the Wizard progressing. Might have a dig into that later. 

I suppose I could manually create the record, trigger the processing, and then perhaps redirect the user to the edit page, all within the afterValidation() method? hmm...

**Reply by dave-mills** [reply]
Ah! Got it. I was getting stuck trying to use the same methods that Filament uses to create new records. But of course I can just do it manually! And returning a redirect works fine. So:

```php
Step::make('1. Xlsform')
    ->description('Upload your XLSForm file and give it a title')
    ->schema(
        XlsformTemplateResource::getCreateFields(),
    )
->afterValidation(function(Get $get) {

    // manually create the new record.
    // no need for validation (we've already validated)
    $xlsformTemplate = XlsformTemplate::create([
        'title' => $get('title'),
    ]);

    $files = $get('xlsfile');

    // the xlsfile upload field is limited to 1 file, but it still returns an array. So get the first one and add to the model. (Note, I'm using Spatie Media Library)
    $xlsformTemplate->addMedia(collect($files)->first())->toMediaCollection('xlsform_file');

    // do processing that normally is triggered on afterCreate()
    $this->processRecord($xlsformTemplate);

    // returning a redirect here will take the user to the edit page for the created record.
    return redirect($this->getResource()::getUrl('edit', ['record' => $xlsformTemplate]));

}),
```


Thanks for the point in the right direction!
